playwright-tests/utils/template-server.ts

```diff
@@ -0,0 +1,286 @@
+      )
+      .map((entry: any) => entry.name);
+
+    for (const templateDir of templateDirs) {
+      const templatePath = join(templatesRoot, templateDir);
+      const packageJsonPath = join(templatePath, "package.json");
+
+      if (fs.existsSync(packageJsonPath)) {
+        try {
+          const packageJson = JSON.parse(
+            fs.readFileSync(packageJsonPath, "utf8"),
+          );
+          const template = this.analyzeTemplate(
+            templateDir,
+            templatePath,
+            packageJson,
+          );
+          if (template) {
+            this.templates.push(template);
+          }
+        } catch (error) {
+          console.warn(
+            `Failed to parse package.json for ${templateDir}:`,
+            error,
+          );
+        }
+      }
+    }
+  }
+
+  private analyzeTemplate(
+    name: string,
+    path: string,
+    packageJson: any,
+  ): Template | null {
+    const scripts = packageJson.scripts || {};
+    const dependencies = {
+      ...packageJson.dependencies,
+      ...packageJson.devDependencies,
+    };
+
+    if (!scripts.dev) {
+      console.warn(`Template ${name} has no dev script, skipping`);
+      return null;
+    }
+
+    // Determine framework and default port
+    let framework: Template["framework"] = "wrangler";
+    let port = 8787; // Default wrangler port
+
+    if (dependencies["vite"] || scripts.dev.includes("vite")) {
+      framework = "vite";
+      port = 5173;
+    } else if (dependencies["next"] || scripts.dev.includes("next")) {
+      framework = "next";
+      port = 3000;
+    } else if (dependencies["astro"] || scripts.dev.includes("astro")) {
+      framework = "astro";
+      port = 4321;
+    } else if (
+      dependencies["@remix-run/dev"] ||
+      scripts.dev.includes("remix")
+    ) {
+      framework = "remix";
+      port = 5173;
+    } else if (
+      dependencies["@react-router/dev"] ||
+      scripts.dev.includes("react-router")
+    ) {
+      framework = "react-router";
+      port = 5173;
+    }
+
+    return {
+      name,
+      path,
+      port,
+      devCommand: scripts.dev,
+      framework,
+    };
+  }
+
+  async startServer(templateName: string): Promise<string> {
+    const template = this.templates.find((t) => t.name === templateName);
+    if (!template) {
+      throw new Error(`Template ${templateName} not found`);
+    }
+
+    // If using live URLs, return the live URL directly
+    if (this.useLiveUrls) {
+      const liveUrl = this.getLiveUrl(templateName);
+      console.log(`Using live URL for ${templateName}: ${liveUrl}`);
+      return liveUrl;
+    }
+
+    if (this.servers.has(templateName)) {
+      console.log(`Server for ${templateName} already running`);
+      return `http://localhost:${template.port}`;
+    }
+
+    console.log(
+      `Starting server for ${template.name} on port ${template.port}...`,
+    );
+
+    const server = spawn("npm", ["run", "dev"], {
```

Comment:
Should we add `dev` as a required script in our linter? Or just skip templates that don't have one?